src/proposal.mligo

```diff
@@ -341,6 +365,65 @@ let delete_proposal
     Set.remove (start_date, proposal_key) store.proposal_key_list_sort_by_date
   }
 
+[@inline]
+let fraction_to_quorum_fraction(n, d : nat * nat): unsigned_quorum_fraction
+  = { numerator = (n * quorum_denominator) / d }
+
+[@inline]
+let to_signed(n : unsigned_quorum_fraction): quorum_fraction
+  = { numerator = int(n.numerator) }
+
+[@inline]
+let to_unsigned(n : quorum_fraction): unsigned_quorum_fraction
+  = { numerator = match is_nat(n.numerator) with
+              | Some n -> n
+              | None -> (failwith("BAD_UNSIGNED_CONVERSION"):nat)
+    }
+
+let update_quorum(store, config : storage * config): storage =
+  let current_period = get_current_period_num(store.start_time, config.voting_period) in
+  if store.quorum_threshold_at_stage.last_updated_period = current_period
+    then store // Quorum has been updated in this period, so no change is required.
+    else
+      match is_nat(current_period - store.period_staked.last_updated_period) with
+        | Some elapsed -> if elapsed >= 2n
```

Comment:
As I was saying I suggest not relying on the `last_updated_period` to always be a proposing stage, instead I think it is more robust to allow `voting` stages as well and here do:
- if `elapsed >= 2n` or `elapsed = 1n && is_proposing_stage(current_period)` then update the `quorum_threshold_at_stage`
- otherwise if `elapsed = 1n` only update the `last_updated_period`

What do you think?

Reply:
I have changed them to count the cycles, calculated as `cycle = (period + 1) /2`

So period `0` should be cycle `0`, so it should be a valid value for initilization. Then both staked tokens and quorum threhold are tracked for cycles. In effect AFAIU it amounts to the same calculation that you do here.